### PR TITLE
Store the profiled information on the node

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The daemon now sends the profiled information as part of the `CONNECTED` message. This needs to be stored on the `Node` object.

Currently it is theoretically possible for a `node_name` to be in multiple partitions. However it is unclear whether these are indeed the same not or not. Dealing with this issue is out of scope of this PR, so instead only the default partition is considered.

The newly profiled information has been added to the serializer.

### CODE REVIEW

**Comment by benarmston** [comment]
> Currently it is theoretically possible for a `node_name` to be in multiple partitions. However it is unclear whether these are indeed the same not or not. Dealing with this issue is out of scope of this PR, so instead only the default partition is considered.

Nodes can be in multiple partitions.  Consider the following, there is a partition "compute" containing any nodes named `nodeXX`, a partition `highmem` containing any nodes named `highmemXX` and a partition "all" containing all nodes.  Any node in the partition "compute" is also in the partition "all".  Similar for nodes in the partition, "highmem".

I'm not certain that that is a useful feature, but it is what the data model supports and what was intended when I wrote that part of the data model.  It is also the behaviour supported by slurm.  Here is example output from slurm:

```
[flight@gateway1 (ben2) ~]$ sinfo
PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST 
all*         up   infinite      2  down* node[01-02] 
compute      up   infinite      2  down* node[01-02] 
[flight@gateway1 (ben2) ~]$ sinfo  --Format Partition,NodeHost,CPUs,Memory
PARTITION           HOSTNAMES           CPUS                MEMORY              
all*                node01              1                   1                   
all*                node02              1                   1                   
compute             node01              1                   1                   
compute             node02              1                   1                   
[flight@gateway1 (ben2) ~]$ sinfo  --Format NodeHost,CPUs,Memory
HOSTNAMES           CPUS                MEMORY              
node01              1                   1                   
node02              1                   1                   
[flight@gateway1 (ben2) ~]$ 
```

---

EDIT:

I see now that when partitions have duplicate `node_name`s they actually have different `Node` instances.  That is a bug that made no difference until now.

File: app/websocket_app.rb [hunk 1]
```diff
@@ -272,33 +275,55 @@ class WebsocketApp
...
+
   def call(env)
     Async::WebSocket::Adapters::Rack.open(env) do |connection|
       begin
         message = connection.read
         unless message.is_a?(Hash) && message[:command] == 'CONNECTED'
           Async.logger.info("Badly formed connection message #{message.inspect}")
           connection.close
-          break
+          next
```
Comment (benarmston) [inline app/websocket_app.rb:284]: Why the change here?

File: app/websocket_app.rb [hunk 2]
```diff
@@ -272,33 +275,55 @@ class WebsocketApp
...
+    end
+    return unless node
+
+    attributes = node.attributes.dup
+    attributes.cpus   = message[:cpus]    if message.key?(:cpus)
+    attributes.gpus   = message[:gpus]    if message.key?(:gpus)
+    attributes.memory = message[:memory]  if message.key?(:memory)
+
+    if attributes.valid?
+      node.attributes = attributes
```
Comment (benarmston) [inline app/websocket_app.rb:268]: Could we get some logging either here or in the node connected log to show the attributes that have been set for it?

**Review by benarmston — CHANGES_REQUESTED** [review]
This looks close.  I'd like some logging to be added when the node's attributes are updated.

We also need to address the multiple nodes with the same name issue.  Can you add a `NodeRegistry` or some other means so that `Configuration#partitions=` can lookup the nodes rather than creating a new one each time.    Likewise `WebsocketApp#update_node_attributes` would use the same lookup mechanism so that nodes that aren't in the default partition can be updated too.

File: app/websocket_app.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply app/websocket_app.rb:284]: I was experiencing a `LocalJumpError` when a malformed packet was being sent, which is why it went undetected for so long.

I'm not entirely sure how the internals of `Async::WebSocket::Adapters::Rack` work, but logically it makes sense to use `next` instead of `break`. In the event of an error, we want the adaptor to move to the next connection instead of breaking the process loop entirely.

Comment (benarmston) [reply app/websocket_app.rb:284]: So it looks like there is no loop in the internals of `Async::WebSocket::Adapters::Rack`.  However there is a call to the provided block (`block.call`).  The `next` causes block to terminate early, whereas `break` results in an error due to the lack of a look.

Good catch @WilliamMcCumstie

File: app/websocket_app.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply app/websocket_app.rb:268]: Logging added

**Comment by benarmston** [comment]
We're still only updating a single `Node` instance.  Can you add the `NodeRegistry`, and change `Configuration#partitions=` and `WebsocketApp#update_node_attributes` accordingly?

**Comment by WilliamMcCumstie** [comment]
Rebased to resolve conflicts and added the `NodeRegistry`

**Comment by benarmston** [comment]
This looks good now.
